ibis/expr/deferred.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+import functools
+import operator
+from typing import TYPE_CHECKING, Any, Callable, TypeVar
+
+import toolz
+
+if TYPE_CHECKING:
+    import ibis.expr.types as ir
+
+
+class Deferred:
```

Comment:
We can use `Annotable` here as well.

Reply:
I'm not sure we get much by doing that here, and it looks to require use of `rlz` which introduces a circular import.

Reply:
Indeed, perhaps validate that `resolve` is callable then?

Reply:
`resolve` should always be callable, and isn't user facing, so I'll add an `assert`.

Reply:
Done.